src/ui/Frame.cpp

```diff
@@ -0,0 +1,1164 @@
+	for (auto slider : sliders) {
+		slider->draw(_size, scroll);
+	}
+
+	// draw subframes
+	for (auto frame : frames) {
+		frame->draw(_size, scroll);
+	}
+
+	// root frame draws tooltip
+	if (!parent) {
+		if (tooltip && tooltip[0] != '\0') {
+			Font* font = Font::get(tooltip_text_font);
+			if (font) {
+				Text* text = Text::get(tooltip, font->getName());
+				SDL_Rect src;
+				src.x = mousex + 20 * ((float)Frame::virtualScreenX / xres);
+				src.y = mousey;
+				src.w = text->getWidth() + 2;
+				src.h = text->getHeight() + 2;
+
+				int border = tooltip_border_width;
+
+				drawRect(&src, tooltip_border_color, (Uint8)(tooltip_border_color>>mainsurface->format->Ashift));
+
+				SDL_Rect src2{src.x + border, src.y + border, src.w - border * 2, src.h - border * 2};
+				drawRect(&src2, tooltip_background, (Uint8)(tooltip_background>>mainsurface->format->Ashift));
+
+				text->drawColor(SDL_Rect{0,0,0,0}, SDL_Rect{src.x + 1, src.y + 1, 0, 0}, tooltip_text_color);
+			}
+		}
+	}
+}
+
+Frame::result_t Frame::process() {
+	result_t result = process(size, actualSize, true);
+
+	tooltip = nullptr;
+	if (result.tooltip && result.tooltip[0] != '\0') {
+		if (SDL_GetTicks() - result.highlightTime >= tooltipTime) {
+			tooltip = result.tooltip;
+		}
+	}
+	postprocess();
+
+	return result;
+}
+
+Frame::result_t Frame::process(SDL_Rect _size, SDL_Rect _actualSize, bool usable) {
+	result_t result;
+	result.removed = false;
+	result.usable = usable;
+	result.highlightTime = SDL_GetTicks();
+	result.tooltip = nullptr;
+
+	if (disabled) {
+		return result;
+	}
+	if (shootmode) {
+		return result;
+	}
+
+	_size.x += std::max(0, size.x - _actualSize.x);
+	_size.y += std::max(0, size.y - _actualSize.y);
+	if (size.h < actualSize.h) {
+		_size.w = std::min(size.w - sliderSize, _size.w - sliderSize - size.x + _actualSize.x) + std::min(0, size.x - _actualSize.x);
+	} else {
+		_size.w = std::min(size.w, _size.w - size.x + _actualSize.x) + std::min(0, size.x - _actualSize.x);
+	}
+	if (size.w < actualSize.w) {
+		_size.h = std::min(size.h - sliderSize, _size.h - sliderSize - size.y + _actualSize.y) + std::min(0, size.y - _actualSize.y);
+	} else {
+		_size.h = std::min(size.h, _size.h - size.y + _actualSize.y) + std::min(0, size.y - _actualSize.y);
+	}
+	if (_size.w <= 0 || _size.h <= 0)
+		return result;
+
+	SDL_Rect fullSize = _size;
+	fullSize.h += (actualSize.w > size.w) ? sliderSize : 0;
+	fullSize.w += (actualSize.h > size.h) ? sliderSize : 0;
+
+	Sint32 mousex = (mousex / (float)xres) * (float)Frame::virtualScreenX;
+	Sint32 mousey = (mousey / (float)yres) * (float)Frame::virtualScreenY;
+	Sint32 omousex = (omousex / (float)xres) * (float)Frame::virtualScreenX;
+	Sint32 omousey = (omousey / (float)yres) * (float)Frame::virtualScreenY;
+
+	if (selected) {
+		// unselect list
+		if (keystatus[SDL_SCANCODE_ESCAPE]) {
+			keystatus[SDL_SCANCODE_ESCAPE] = 0;
+			deselect();
+			if (!widgetBack.empty()) {
+				Frame* root = findSearchRoot(); assert(root);
```

Comment:
Ah, are we building with asserts?

Reply:
I'm very fond of asserts as it explicitly draws the line between a working and broken state. there is no conceivable case in this code where you'd want root to be null. the assert implies that will never happens at runtime

Reply:
Hmm, ok. I hope it never happens at runtime :)

Reply:
(I usually just do checks for conditions over asserts, but even when I use asserts, I leave in some error handling code after the assert for the production builds...)